server.js

```diff
@@ -13,8 +14,126 @@ app.get("/", function (request, response) {
   response.send("Hotel booking server.  Ask for /bookings, etc.");
 });
 
-// TODO add your routes and helper functions here
+//reading all bookings
+app.get("/bookings", (req, res) => {
+  res.status(200).send(bookings);
+});
+
+app.post("/bookings", (req, res) => {
+  const requestBody = req.body;
+
+  const newBooking = {
+    id: requestBody.id,
+    title: requestBody.title,
+    firstName: requestBody.firstName,
+    surname: requestBody.surname,
+    email: requestBody.email,
+    roomId: requestBody.roomId,
+    checkInDate: requestBody.checkInDate,
+    checkOutDate: requestBody.checkOutDate,
+  };
+
+  // validation
+
+  const validBooking =
+    (!!requestBody.id || requestBody.id === 0) &&
+    !!requestBody.title &&
+    !!requestBody.firstName &&
+    !!requestBody.surname &&
+    !!requestBody.email &&
+    emailValidation(requestBody.email) &&
+    !!requestBody.roomId &&
+    !!requestBody.checkInDate &&
+    !!requestBody.checkOutDate &&
+    checkInAndOutDatedValidation(
+      requestBody.checkInDate,
+      requestBody.checkOutDate
+    );
+
+  if (validBooking) {
+    bookings.push(newBooking);
+    res.status(200).send(bookings);
+  } else {
+    res.status(400).send("Some information is missing");
+  }
+});
+
+// Read by Id
+app.get("/bookings/:id", (req, res) => {
+  const bookingId = +req.params.id;
+  const searchedBookingById = bookings.filter(
+    (booking) => booking.id === bookingId
+  );
+  res.status(200).send(searchedBookingById);
+});
+
+//Deleting by id
+app.delete("/bookings/:id", (req, res) => {
+  const bookingId = +req.params.id;
+  const bookingIndex = bookings.findIndex(
+    (booking) => booking.id === bookingId
+  );
+
+  bookings.splice(bookingIndex, 1);
+
+  res.status(200).send(bookings);
+});
+
+//Level 3 search by date
+
+// convet date to numbers
+const dateToNumber = (date) => Number(date.replaceAll("-", ""));
+
+app.get("/bookings/search/:date", (req, res) => {
+  const date = dateToNumber(req.params.date);
+
+  const bookingByDate = bookings.filter((booking) => {
+    // this is filtering out if the date is * earlier then the checkout and later then the checkin
+    return (
+      dateToNumber(booking.checkInDate) <= date &&
+      dateToNumber(booking.checkOutDate) >= date
+    );
+  });
+
+  if (bookingByDate) {
+    res.status(200).send(bookingByDate);
+  } else {
+    res.status(400).send("No booking has this date");
+  }
+});
+
+// level 4
+
+const emailValidation = (email) => {
+  email.includes("@") ? email : "Not valid email";
```

Comment:
Lot more to email validation than this I'm afraid, worth reading up/using a libary. Also, it would be best to return a boolean from this function. At the moment, this function always returns a non empty string so it will always evaluate to true.